segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeperstorage/LogStorageManager.java

```diff
@@ -0,0 +1,817 @@
+                   /** Delete metadata and cache for the source once the operation is complete. */
+                   .thenCompose(v -> this.zkClient.delete()
+                                                  .withOptions(new HashSet<>(
+                                                          Arrays.asList(DeleteOption.guaranteed,
+                                                                  DeleteOption.deletingChildrenIfNeeded)))
+                                                  .forPath(getZkPath(sourceSegment))
+                                                  .toCompletableFuture()
+                                                  .thenApply(ret -> {
+                                                      ledgers.remove(sourceSegment);
+                                                      return null;
+                                                  }));
+    }
+
+    /**
+     * Deletes a segment along with all the data and metadata.
+     * @param segmentName name of the segment to be deleted.
+     * @return A CompletableFuture which completes once the delete is complete.
+     */
+    public CompletableFuture<Void> delete(String segmentName) {
+        return this.getOrRetrieveStorageLedger(segmentName, false)
+                   .thenCompose(ledger -> {
+                       if (ledger.getContainerEpoch() > this.containerEpoch) {
+                           throw new CompletionException(new StorageNotPrimaryException(segmentName));
+                       }
+                       return this.zkClient.delete()
+                                           .withOptions(new HashSet<>(
+                                                   Arrays.asList(DeleteOption.guaranteed,
+                                                           DeleteOption.deletingChildrenIfNeeded)))
+                                           .forPath(getZkPath(segmentName))
+                                           .toCompletableFuture()
+                                           .thenApply(v -> {
+                                               ledgers.remove(segmentName);
+                                               return ledger;
+                                           });
+                   })
+                   .thenCompose(ledger -> ledger.deleteAllLedgers());
+    }
+
+    //endregion
+
+
+    public CompletableFuture<LedgerData> createLedgerAt(String streamSegmentName, int offset) {
+        CompletableFuture<LedgerData> retVal = new CompletableFuture<>();
+
+        bookkeeper.asyncCreateLedger(config.getBkEnsembleSize(),
+                config.getBkWriteQuorumSize(), LEDGER_DIGEST_TYPE, config.getBKPassword(),
+                (errorCode, ledgerHandle, future) -> {
+                    if (errorCode == 0) {
+                        LedgerData lh = new LedgerData(ledgerHandle, offset, 0, this.containerEpoch);
+                        zkClient.create().forPath(getZkPath(streamSegmentName) + "/" + offset, lh.serialize())
```

Comment:
Use `Futures.completeAfter` here. This makes the code cleaner and ensures that `retVal` is completed even in case of sync failure